src/java.base/share/classes/jdk/internal/classfile/impl/ClassFileImpl.java

```diff
@@ -135,6 +141,35 @@ public ClassModel parse(byte[] bytes) {
         return new ClassImpl(bytes, this);
     }
 
+    @Override
+    public ClassModel parse(final MemorySegment bytes) {
+        AbstractMemorySegmentImpl amsi = (AbstractMemorySegmentImpl) bytes;
+        if (amsi.unsafeGetBase() instanceof byte[] ba) {
+            if (amsi.unsafeGetOffset() == 0 && amsi.byteSize() == ba.length) {
+                return parse(ba);
+            }
+        }
+        // must copy the bytes
+        return parse(bytes.toArray(ValueLayout.JAVA_BYTE));
+    }
+
+    @Override
+    public ClassModel parse(final ByteBuffer bytes) {
+        // defensive copy
+        ByteBuffer dup = bytes.duplicate();
```

Comment:
The array version has no defensive copy. Doesn't this defeat the performance goal of the API?

Reply:
No; duplicating a buffer does not duplicate its contents, only the wrapping object.